src/controller/SetUpCodePairer.cpp

```diff
@@ -40,61 +41,44 @@ using namespace chip::Tracing;
 namespace chip {
 namespace Controller {
 
-namespace {
-
-CHIP_ERROR GetPayload(const char * setUpCode, SetupPayload & payload)
-{
-    bool isQRCode = strncmp(setUpCode, kQRCodePrefix, strlen(kQRCodePrefix)) == 0;
-    if (isQRCode)
-    {
-        ReturnErrorOnFailure(QRCodeSetupPayloadParser(setUpCode).populatePayload(payload));
-        VerifyOrReturnError(payload.isValidQRCodePayload(), CHIP_ERROR_INVALID_ARGUMENT);
-    }
-    else
-    {
-        ReturnErrorOnFailure(ManualSetupPayloadParser(setUpCode).populatePayload(payload));
-        VerifyOrReturnError(payload.isValidManualCode(), CHIP_ERROR_INVALID_ARGUMENT);
-    }
-
-    return CHIP_NO_ERROR;
-}
-} // namespace
-
 CHIP_ERROR SetUpCodePairer::PairDevice(NodeId remoteId, const char * setUpCode, SetupCodePairerBehaviour commission,
                                        DiscoveryType discoveryType, Optional<Dnssd::CommonResolutionData> resolutionData)
 {
     VerifyOrReturnErrorWithMetric(kMetricSetupCodePairerPairDevice, mSystemLayer != nullptr, CHIP_ERROR_INCORRECT_STATE);
     VerifyOrReturnErrorWithMetric(kMetricSetupCodePairerPairDevice, remoteId != kUndefinedNodeId, CHIP_ERROR_INVALID_ARGUMENT);
 
-    SetupPayload payload;
-    ReturnErrorOnFailure(GetPayload(setUpCode, payload));
+    std::vector<SetupPayload> payloads;
+    ReturnErrorOnFailure(SetupPayload::FromStringRepresentation(setUpCode, payloads));
 
-    if (resolutionData.HasValue())
+    if (resolutionData.HasValue() && payloads.size() == 1 && mSetupPayloads.size() == 1)
```

Comment:
some explanation comment on why the `== 1` is happening would be nice: when do we expect this codepath to be taken and when not (i.e. when payloads has 1 and not and when is mSetuppayload 1 or not).

Reply:
And in particular what does the code below do? It seems to do a check of "if all inputs are the same as what is already stored in the object, then call discovered" ... why does that check not extend to "if I pass an identical payload array" and only applies to "if I send a single identical element"?

Reply:
This is just preserving an existing hacky codepath that only ever has one value.  I'll add comments.

Reply:
I guess my question is about why is a single element special. Hack code path seemed to have been "if input was identical" so why does that not extend to "if multi-item input is identical".

I do not find the existing code too readable :/

Reply:
Because the whole premise of the hack codepath is that we have _a single resolutionData_, which therefore represents only one thing.  In practice, this is only used by chip-tool for things that don't involve concatenated QR codes.

I agree it's not very readable; I will add documentation.